### PR TITLE
Skill Tactic Play Initial Restructure

### PR DESCRIPTION
**Pull request by RobotJente**
**Summary**
In this PR I change the structure of the decision-making process for our AI. The new structure has Plays, which are sequences of robot skills to accomplish a specific task. Each play has a set of invariants and preconditions.

**Which play to choose is decided as follows:**
First, plays are filtered out if their preconditions are false. Then, out of the remaining plays, a Play Decider module makes a decision on which play should be chosen. Finally, this play is executed. Currently the play execution is a behaviour tree that is tailor made for this situation. In this PR I've tested the structure with 2 plays: a Pass and Play Play, and a Dummy Play. In the pass and play, robot 1 passes the ball to robot 2. Robot 2, after receiving the ball, will then shoot it at a location determined beforehand (should be done according to some heatmap, but currently hardcoded to (2,2)). Meanwhile, robot 1 is already driving to the location (2,2). 

Each tick, it is checked whether a play is still valid for execution by checking if all of its invariants are true. If this is not the case, then a new play is selected according to the process described above. One potential issue with this structure is that invariants suddenly become false right after the play is chosen, and you could get oscillatory behaviour. A solution I have seen in other teams is to make sure the play is at least executed for a set amount of time. CMDragons calls this "Hysterisis" in their <a href = "https://www.cs.cmu.edu/~mmv/papers/05STP-IEEE.pdf">paper</a> (@rolfvdhulst if you have not yet read this paper, you would really like it!). 

The hardest part is defining proper abstractions for the skills and the plays (CMD's tactics are basically our skills right now), but that should not be the purpose of this PR as this mainly restructures what we have now. 

I would also like to mention that whilst the place where the plays is decided is the PlayDecider, the plays also score themselves. I am not a huge fan of this, but it's easier to implement and that's why I did it this way. However, it gives a lot of responsibility to the Plays themselves, which is not what I want. I don't see a good solution for this right now, but I am open to suggestions :)

This is still under construction, but I am opening this PR so people can provide feedback if they would like to. Something somewhere is also segfaulting but it only happens sometimes, so I will play segfault detective for a bit more.

### CODE REVIEW

**Comment by rolfvdhulst** [comment]
>  CMDragons calls this "Hysterisis" in their paper (@rolfvdhulst if you have not yet read this paper, you would really like it!).

Hysteresis is a generally known scientific concept-> Essentially you say that the state of a system is dependent on it's history.  In practice I know it from the 'thermometer example' where you say e.g. that a heater only turns on if temperature is below e.g. 18 degrees but turns off if the temperature is above 20 degrees.

File: include/roboteam_ai/ApplicationManager.h [hunk 1]
```diff
@@ -36,6 +37,7 @@ class ApplicationManager {
     void updateTrees();
     bt::Node::Status runStrategyTree();
     void runKeeperTree();
+    void decidePlay(ai::world::World* world, ai::world::Field* field);
```
Comment (Yuhanun) [inline include/roboteam_ai/ApplicationManager.h:40]: Why are these mutable pointers?

File: include/roboteam_ai/analysis/PlayChecker.h [hunk 2]
```diff
@@ -12,13 +12,22 @@ namespace rtt::ai::analysis {
...
+         */
+        PlayChecker();
         /**
          * Updates the PlayChecker. When this function is called, we check if the play is still valid for the current gamestate,
          * and if it is not new plays are calculated that are valid for the gamestate.
          */
-        void update(world::World *world, world::Field *field);
+        bool update(world::World *world, world::Field *field);
 
+        const std::vector<std::shared_ptr<rtt::ai::analysis::Play>> &getValidPlays() const;
```
Comment (Yuhanun) [inline include/roboteam_ai/analysis/PlayChecker.h:28]: Why shared_ptr?

File: include/roboteam_ai/analysis/PlayChecker.h [hunk 3]
```diff
@@ -31,12 +40,12 @@ namespace rtt::ai::analysis {
         /**
          * Vector of all plays (before pruning)
          */
-        std::vector<Play> allPlays;
+        std::vector<std::shared_ptr<rtt::ai::analysis::Play>> allPlays;
```
Comment (Yuhanun) [inline include/roboteam_ai/analysis/PlayChecker.h:43]: p sure you don't have to `rtt::ai::`

File: include/roboteam_ai/analysis/PlaysObjects/DummyPlay.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+#ifndef RTT_DUMMYPLAY_H
+#define RTT_DUMMYPLAY_H
+
+#include <include/roboteam_ai/utilities/RobotDealer.h>
+#include "analysis/PlaysObjects/Play.h"
+
+namespace rtt::ai::analysis {
+    class DummyPlay : public Play {
+    public:
+        DummyPlay(std::string name);
```
Comment (Yuhanun) [inline include/roboteam_ai/analysis/PlaysObjects/DummyPlay.h:14]: string_view

File: include/roboteam_ai/analysis/PlaysObjects/PassAndPlayPlay.h [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+         /**
+          * Decides if the play is valid to keep playing for the given world
+          * @param world the current world
+          * @param field the current field
+          * @return true if the play is valid, false otherwise
+          */
+        virtual bool isValidPlayToKeep(rtt::ai::world::World* world, rtt::ai::world::Field* field);
+
+        virtual bool isValidPlayToStart(world::World *world, world::Field *field);
```
Comment (Yuhanun) [inline include/roboteam_ai/analysis/PlaysObjects/PassAndPlayPlay.h:35]: Docs?

File: include/roboteam_ai/analysis/PlaysObjects/Play.h [hunk 6]
```diff
@@ -11,34 +11,51 @@
...
     /**
-     * The play has a vector of invariants, when the invariants are false the play is abandoned.
+     * The play has some invariants, defined in the isValidPlayToKeep function. When the invariants are false the play is abandoned.
      *
-     * TODO: Should this object have a behaviourtree associated to it? I think yes
      */
     class Play {
     public:
-        Play();
+        Play(std::string name);
```
Comment (Yuhanun) [inline include/roboteam_ai/analysis/PlaysObjects/Play.h:19]: string_view

File: include/roboteam_ai/analysis/PlaysObjects/Play.h [hunk 7]
```diff
@@ -11,34 +11,51 @@
 
 namespace rtt::ai::analysis {
     /**
-     * The play has a vector of invariants, when the invariants are false the play is abandoned.
+     * The play has some invariants, defined in the isValidPlayToKeep function. When the invariants are false the play is abandoned.
      *
-     * TODO: Should this object have a behaviourtree associated to it? I think yes
      */
     class Play {
```
Comment (Yuhanun) [inline include/roboteam_ai/analysis/PlaysObjects/Play.h:17]: I don't think a virtual abstract class is the way to deal with plays but it's definitely a maintainable OOP approach

File: include/roboteam_ai/bt/BehaviorTree.hpp [hunk 8]
```diff
@@ -33,8 +33,14 @@ class BehaviorTree : public Node {
         void SetGlobalBlackboard(const Blackboard::Ptr &globalBB);
 
     private:
+        std::shared_ptr<BehaviorTree> next;
```
Comment (Yuhanun) [inline include/roboteam_ai/bt/BehaviorTree.hpp:36]: Wait so we have a linked list of behavior trees now?...

File: include/roboteam_ai/bt/BehaviorTree.hpp [hunk 9]
```diff
@@ -33,8 +33,14 @@ class BehaviorTree : public Node {
...
+public:
+    const std::shared_ptr<BehaviorTree> &getNext() const;
```
Comment (Yuhanun) [inline include/roboteam_ai/bt/BehaviorTree.hpp:38]: dont return references to shared_ptr's, defeats their whole purpose

File: include/roboteam_ai/skills/NewPass.h [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+protected:
+    enum PassType {
+        DEFAULT,
+        DEFENSIVE,
+        FREEKICK
+    };
+
+    PassType passType = DEFAULT;
+
+    PassType stringToType(const std::string& type);
```
Comment (Yuhanun) [inline include/roboteam_ai/skills/NewPass.h:29]: string_view would be nicer here

File: include/roboteam_ai/skills/NewPass.h [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+    const double CLOSE_ENOUGH_TO_BALL = 0.7;
+    const double SUCCESSFUL_PASS_ANGLE = 0.6;
```
Comment (Yuhanun) [inline include/roboteam_ai/skills/NewPass.h:32]: constexpr static double ... = ...;

File: src/ApplicationManager.cpp [hunk 12]
```diff
@@ -176,4 +180,17 @@ void ApplicationManager::notifyTreeStatus(bt::Node::Status status) {
...
+    void ApplicationManager::decidePlay(ai::world::World *world, ai::world::Field *field) {
+        bool stillValidPlay = playcheck.update(world, field);
+        if (!stillValidPlay) {
+            auto bestplay = playdecide.decideBestPlay(world, field, playcheck.getValidPlays());
+            rtt::ai::interface::TreeVisualizerWidget::setHasCorrectTree(false);
+            BTFactory::setCurrentTree(bestplay);
+            playcheck.setCurrentPlay(bestplay);
+        }
+        else {
+        }
```
Comment (Yuhanun) [inline src/ApplicationManager.cpp:192]: empty conditional

File: src/analysis/PlayChecker.cpp [hunk 13]
```diff
@@ -12,31 +14,54 @@
...
 
-    /**
-     * Determines what plays are viable given the current world, ref states and invariants/preconditions, and stores them in the validPlays vector
-     * TODO: add lambda here, to make it faster and cleaner
-     */
-    void PlayChecker::determineNewPlays(rtt::ai::world::World* world, rtt::ai::world::Field* field) {
+
+    bool PlayChecker::checkCurrentGameInvariants(rtt::ai::world::World *world, rtt::ai::world::Field *field) {
+        currentPlay->isValidPlayToKeep(world, field);
+    }
```
Comment (Yuhanun) [inline src/analysis/PlayChecker.cpp:27]: No return statement, rip ud2 instruction

File: src/analysis/PlaysObjects/DummyPlay.cpp [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+#include "include/roboteam_ai/analysis/PlaysObjects/DummyPlay.h"
+//
+// Created by jessevw on 14.01.20.
+//
+
+#include <include/roboteam_ai/bt/Role.h>
+#include <include/roboteam_ai/skills/Halt.h>
+#include <include/roboteam_ai/skills/Attack.h>
+#include <include/roboteam_ai/bt/Node.hpp>
+#include <include/roboteam_ai/analysis/PlaysObjects/Invariants/BallOnOurSideInvariant.h>
```
Comment (Yuhanun) [inline src/analysis/PlaysObjects/DummyPlay.cpp:14]: Inconsistent includes

File: CMakeLists.txt [hunk 15]
```diff
@@ -144,6 +149,9 @@ set(SKILLS_SOURCES
         ${PROJECT_SOURCE_DIR}/src/skills/ReflectKick.cpp
         ${PROJECT_SOURCE_DIR}/src/skills/ChipForward.cpp
 
+        # Plays skills sources
```
Comment (mrlukasbos) [inline CMakeLists.txt:152]: plays skills?

File: include/roboteam_ai/ApplicationManager.h [hunk 16]
```diff
@@ -11,12 +11,13 @@
...
 #include "io/IOManager.h"
 #include "treeinterp/BTFactory.h"
-
+#include "analysis/PlaysObjects/PlayDecider.h"
 namespace rtt {
 
 class ApplicationManager {
 private:
     rtt::ai::analysis::PlayChecker playcheck;
+    rtt::ai::analysis::PlayDecider playdecide;
```
Comment (mrlukasbos) [inline include/roboteam_ai/ApplicationManager.h:20]: typo

File: include/roboteam_ai/ApplicationManager.h (hunk 1 above)
Comment (mrlukasbos) [inline include/roboteam_ai/ApplicationManager.h:40]: Why raw pointers?

File: src/analysis/PlaysObjects/DummyPlay.cpp [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+
+    void DummyPlay::moveToNextTactic() {
+        if (tree->getStatus() == bt::BehaviorTree::Status::Success && tree->getNext()) {
+            std::cout << "moving to next tactic!" << std::endl;
+            tree = tree->getNext();
+        }
+    }
+
+    int DummyPlay::scorePlay(rtt::ai::world::World *world, rtt::ai::world::Field *field) {
+        int score = 3;
```
Comment (Yuhanun) [inline src/analysis/PlaysObjects/DummyPlay.cpp:80]: nice :joy:

File: include/roboteam_ai/analysis/PlayChecker.h (hunk 2 above)
Comment (mrlukasbos) [inline include/roboteam_ai/analysis/PlayChecker.h:28]: you don't need the name spacing here, you are already in namespace rtt::ai::analysis.

File: src/analysis/PlaysObjects/PassAndPlayPlay.cpp [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+// Created by jessevw on 14.01.20.
+//
+
+#include <include/roboteam_ai/bt/Role.h>
+#include <include/roboteam_ai/skills/Halt.h>
+#include <include/roboteam_ai/skills/Attack.h>
+#include <include/roboteam_ai/bt/Node.hpp>
+#include <include/roboteam_ai/analysis/PlaysObjects/Invariants/BallBelongsToUsInvariant.h>
+#include <include/roboteam_ai/analysis/PlaysObjects/Invariants/BallOnOurSideInvariant.h>
+#include <include/roboteam_ai/bt/composites/Sequence.hpp>
```
Comment (Yuhanun) [inline src/analysis/PlaysObjects/PassAndPlayPlay.cpp:11]: Include style

File: include/roboteam_ai/analysis/PlaysObjects/DummyPlay.h [hunk 19]
```diff
@@ -0,0 +1,31 @@
+//
+// Created by jessevw on 21.01.20.
+//
+
```
Comment (mrlukasbos) [inline include/roboteam_ai/analysis/PlaysObjects/DummyPlay.h:4]: This class misses documentation.

File: src/analysis/PlaysObjects/PassAndPlayPlay.cpp [hunk 20]
```diff
@@ -0,0 +1,164 @@
...
+        std::shared_ptr<bt::DefaultTactic> tactic = std::make_shared<bt::DefaultTactic>("tactic", bb, robots);
+
+        // Creating the roles for all the robots in the tactic:
+        for (int i = 1; i < robots.size(); i++) {
+            auto localb = std::make_shared<bt::Blackboard>();
+            std::string rolename = "o" + std::to_string(i);
+            localb->setString("ROLE", rolename);
+            std::shared_ptr<bt::Role> temprole = std::make_shared<bt::Role>(rolename);
+
+            if (i == 5) {
```
Comment (Yuhanun) [inline src/analysis/PlaysObjects/PassAndPlayPlay.cpp:130]: Magic number

File: src/analysis/PlaysObjects/PassAndPlayPlay.cpp [hunk 21]
```diff
@@ -0,0 +1,164 @@
...
+                auto s = std::make_shared<bt::Sequence>();
+                s->addChild(pass);
+                auto posbb = std::make_shared<bt::Blackboard>();
+                posbb->setString("goToType", "numTrees");
+                posbb->setVector2("targetPos", Vector2(2,2));
+                auto gtp = std::make_shared<rtt::ai::SkillGoToPos>("go to pos 2,2", posbb);
+                s->addChild(gtp);
+                temprole->addChild(s);
+            }
+            else if (i == 2) {
```
Comment (Yuhanun) [inline src/analysis/PlaysObjects/PassAndPlayPlay.cpp:142]: Magic number

File: include/roboteam_ai/analysis/PlaysObjects/Play.h [hunk 22]
```diff
@@ -11,34 +11,51 @@
...
+        virtual int scorePlay(rtt::ai::world::World* world, rtt::ai::world::Field* field) = 0;
 
         /**
-         *
-         * @return true if all the invariants of this strategy are true
+         * See derived class for documentation
+         * @return to the derived class for documentation
          */
-        bool isValidPlay(rtt::ai::world::World* world, rtt::ai::world::Field* field) ;
+        virtual std::shared_ptr<bt::BehaviorTree> getTreeForWorld() = 0;
```
Comment (mrlukasbos) [inline include/roboteam_ai/analysis/PlaysObjects/Play.h:46]: I would expect world as a parameter here

File: include/roboteam_ai/analysis/PlaysObjects/PlayDecider.h [hunk 23]
```diff
@@ -10,8 +10,24 @@
...
+
+        /**
+         * Decides which play to choose out of the valid plays. Selects the play that best fits the world and field state. If you want to
+         * do meta logic, like making the play selection more aggressive or risky, you should do it in this function
+         * @param world the current world state
+         * @param field the current field state
+         * @param validPlays a vector of plays which have been vetted by the PlayDecider and are deemed appropriate for this situaton
+         * @return the play that best fits the world, and field.
+         */
+        std::shared_ptr<rtt::ai::analysis::Play> decideBestPlay(world::World* world, world::Field* field, std::vector<std::shared_ptr<rtt::ai::analysis::Play>> validPlays);
```
Comment (mrlukasbos) [inline include/roboteam_ai/analysis/PlaysObjects/PlayDecider.h:24]: again you don't need rtt::ai::analysis here, you are already there.

File: src/bt/tactics/DefaultTactic.cpp [hunk 24]
```diff
@@ -29,16 +29,24 @@ Node::Status DefaultTactic::update() {
     if (amountToTick > children.size()) {
         amountToTick = children.size();
     }
-
+    bool allSuccess = true;
```
Comment (Yuhanun) [inline src/bt/tactics/DefaultTactic.cpp:32]: To me, using std::all_of would be clearer here

```cpp
std::all_of(children.begin(), children.begin() + i, [&](auto& child) {
    return child->tick(world, field) == Status::Success;
});
```

File: include/roboteam_ai/interface/widgets/TreeVisualizerWidget.h [hunk 25]
```diff
@@ -23,13 +23,13 @@ class TreeVisualizerWidget : public QTreeWidget {
     QColor getColorForStatus(bt::Node::Status status);
     void addRootItem(bt::Node::Ptr parent, QTreeWidgetItem* QParent);
     std::map<QTreeWidgetItem *, bt::Node::Ptr> treeItemMapping;
-    bool hasCorrectTree = false;
+    static bool hasCorrectTree;
```
Comment (mrlukasbos) [inline include/roboteam_ai/interface/widgets/TreeVisualizerWidget.h:26]: Okay I do not recommend doing things like this. Did you know we have two instances of treevisualizerwidget?

File: include/roboteam_ai/interface/widgets/TreeVisualizerWidget.h [hunk 26]
```diff
@@ -23,13 +23,13 @@ class TreeVisualizerWidget : public QTreeWidget {
...
     MainWindow * parent = nullptr;
     unsigned long long mostTicks = 0;
     GameState recentGameState;
 public:
     explicit TreeVisualizerWidget(MainWindow * parent);
-    void setHasCorrectTree(bool hasCorrectTree);
+    static void setHasCorrectTree(bool hasCorrectTree);
```
Comment (mrlukasbos) [inline include/roboteam_ai/interface/widgets/TreeVisualizerWidget.h:32]: idem.

File: include/roboteam_ai/skills/NewPass.h [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+
+    bool forcePass = false;
```
Comment (mrlukasbos) [inline include/roboteam_ai/skills/NewPass.h:34]: I see this is largely based on the previous pass skill. A Pass asks for a 'pass object' that will make your life much more simple.

File: src/analysis/PlaysObjects/PassAndPlayPlay.cpp [hunk 28]
```diff
@@ -0,0 +1,164 @@
...
+    int PassAndPlayPlay::scorePlay(rtt::ai::world::World *world, rtt::ai::world::Field *field) {
+        int score = 0;
+        if (BallBelongsToUsInvariant::isValid(world, field)) {
+            score = 10;
+        }
+        return score;
+    }
+
+
+    void PassAndPlayPlay::makeTree2() {
```
Comment (mrlukasbos) [inline src/analysis/PlaysObjects/PassAndPlayPlay.cpp:72]: makeTree1,2,3? you can make abstractions for this

File: src/analysis/PlaysObjects/PassAndPlayPlay.cpp [hunk 29]
```diff
@@ -0,0 +1,164 @@
...
+        std::shared_ptr<bt::DefaultTactic> tactic = std::make_shared<bt::DefaultTactic>("tree2", bb, robots);
+
+        // Creating the roles for all the robots in the tactic:
+        for (int i = 1; i < robots.size(); i++) {
+            auto localbb = std::make_shared<bt::Blackboard>();
+            std::string rolename = "o" + std::to_string(i);
+            localbb->setString("ROLE", rolename);
+            std::shared_ptr<bt::Role> temprole2 = std::make_shared<bt::Role>(rolename);
+
+            if (i == 5) {
```
Comment (mrlukasbos) [inline src/analysis/PlaysObjects/PassAndPlayPlay.cpp:93]: magic value

File: src/skills/NewPass.cpp [hunk 30]
```diff
@@ -0,0 +1,112 @@
...
+
+    std::cout << "shot properly:" << (hasShot && ballIsMovingFast && ballIsMovingToReceiver) << std::endl;
+    return (hasShot && ballIsMovingFast && ballIsMovingToReceiver);
+}
+
+NewPass::PassType NewPass::stringToType(const std::string& type) {
+    if (type == "defensive") {
+        return DEFENSIVE;
+    } else
+    if (type == "freeKick") {
```
Comment (Yuhanun) [inline src/skills/NewPass.cpp:103]: else
if ?

Did you apply clang-format?

File: src/analysis/PlaysObjects/PassAndPlayPlay.cpp (hunk 20 above)
Comment (mrlukasbos) [inline src/analysis/PlaysObjects/PassAndPlayPlay.cpp:130]: magic value

File: src/skills/KickTo.cpp [hunk 31]
```diff
@@ -13,25 +13,20 @@ KickTo::KickTo(string name, bt::Blackboard::Ptr blackboard)
 }
 void KickTo::onInitialize() {
     std::string type=properties->getString("type");
-    if (type=="shootout"){
-        shootPos = Vector2(field->get_field().get(FIELD_LENGTH) * 0.2, 0); // 2.4 m for A field, 1.8 for B
-    }
-    else{
-        shootPos = Vector2(0,0);
-    }
+    shootPos = properties->getVector2("where");
```
Comment (mrlukasbos) [inline src/skills/KickTo.cpp:16]: what does where mean? a direction? a coordinate? a robot id?

File: src/skills/NewPass.cpp [hunk 32]
```diff
@@ -0,0 +1,112 @@
...
+#include "skills/NewPass.h"
+
+namespace rtt {
+namespace ai {
+
+NewPass::NewPass(string name, bt::Blackboard::Ptr blackboard)
+        :Skill(std::move(name), std::move(blackboard)) { }
+
+void NewPass::onInitialize() {
+    robotToPassToID = - 1;
```
Comment (mrlukasbos) [inline src/skills/NewPass.cpp:23]: If you make a class for Pass you can avoid all of this

File: src/treeinterp/BTFactory.cpp [hunk 33]
```diff
@@ -62,16 +63,9 @@ void BTFactory::makeTrees() {
...
-//    return treefound->second;
-
-if (strategyRepo.find(treeName) != strategyRepo.end()) {
-        return strategyRepo.find(treeName)->second;
-    }
-    std::cerr << "NO STRATEGY BY THAT NAME:" << treeName.c_str() << std::endl;
-    return nullptr;
+    std::cout << play->getName() << " is currently being played" << std::endl;
+    auto tree = BTFactory::play->getTreeForWorld();
+    return tree;
```
Comment (Yuhanun) [inline src/treeinterp/BTFactory.cpp:68]: return BTFactory::play->getTreeForWorld();

**Review by Yuhanun — CHANGES_REQUESTED** [review]

**Review by mrlukasbos — CHANGES_REQUESTED** [review]
The PR is quite big and the changes are large as well. 
I suggest splitting this PR into three or four smaller pull requests. 

- One for the new pass
- One for the dummy play
- One for the PassAndPlay-play
- One for the other small stuff.

**Pull request closed without merging**